src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreMetricsInstrumentationOptions.cs

```diff
@@ -0,0 +1,45 @@
+// <copyright file="AspNetCoreInstrumentationOptions.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Http;
+
+namespace OpenTelemetry.Instrumentation.AspNetCore
+{
+    /// <summary>
+    /// Options for metrics requests instrumentation.
+    /// </summary>
+    public class AspNetCoreMetricsInstrumentationOptions
+    {
+        /// <summary>
+        /// Gets or sets a Filter function that determines whether or not to collect telemetry about requests on a per request basis.
+        /// The Filter gets the HttpContext, and should return a boolean.
+        /// If Filter returns true, the request is collected.
+        /// If Filter returns false or throw exception, the request is filtered out.
+        /// </summary>
+        public Func<HttpContext, bool> Filter { get; set; }
+
+        /// <summary>
+        /// Gets or sets an function to enrich an recorded metric with additional custom tags.
+        /// </summary>
+        /// <remarks>
+        /// <para><see cref="HttpContext"/>: the HttpContext object.</para>
+        /// </remarks>
+        public Func<HttpContext, IEnumerable<KeyValuePair<string, object>>> EnrichWithCustomTags { get; set; }
```

Comment:
* nit: We could probably just call this `Enrich`
* Let's change the signature to be: `Func<HttpContext, System.Diagnostics.TagList>`. The nice thing about `TagList` is it is a struct so this can be done without an allocation for a certain number of tags (8 at the moment, IIRC).

Reply:
Actually on the signature, if you want to achieve max perf do something like this:

```csharp
public class AspNetCoreMetricsInstrumentationOptions
{
        public delegate void AspNetCoreMetricEnrichmentFunc(HttpContext context, out TagList tags);

        public AspNetCoreMetricEnrichmentFunc Enrich { get; set; }
}
```

* `TagList` ~avoids allocations.
* The `out` usage avoids copying a potentially very big `struct`.

PS: In that snippet I nested the delegate in the options class, I think that is probably fine to do for this.

Reply:
Reworked as suggested